### ISSUE TITLE
[Actor-Critic做市策略] | Actor-Critic Market Making

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4491]
**标签**: quantitative-strategy
### 摘要
基于Actor-Critic强化学习框架的做市商策略，自适应优化买卖价差和库存管理，平衡执行成本与订单流信息价值。

### 算法描述

**核心原理**

做市商需要在提供流动性的同时管理库存风险和订单流逆向选择风险。Actor-Critic架构允许策略同时学习报价动作（Actor）和价值估计（Critic），通过持续优化实现收益最大化。

**数学模型**

1. **做市商利润分解**：
   $$\pi = \underbrace{(b - a) \cdot \lambda}_{\text{价差收益}} - \underbrace{\alpha \cdot \sigma^2 \cdot Q^2}_{\text{库存风险成本}} - \underbrace{\beta \cdot |Q| \cdot \Delta P}_{\text{逆向选择损失}}$$
   其中 $b, a$ 为买卖价，$\lambda$ 为订单到达率，$\alpha$ 为库存风险系数，$Q$ 为持仓，$\sigma$ 为价格波动率，$\Delta P$ 为价格变动。

2. **Actor网络（策略）**：
   $$\mu(s) = \text{softmax}(W_2 \cdot \text{ReLU}(W_1 \cdot s + b_1) + b_2)$$
   输出为买卖价差和报价量的概率分布。

3. **Critic网络（价值）**：
   $$V(s) = w_2 \cdot \text{ReLU}(w_1 \cdot s + c_1) + c_2$$
   估计状态价值函数，用于策略梯度计算。

4. **优势函数（Advantage）**：
   $$A(s, a) = Q(s, a) - V(s) = r + \gamma V(s') - V(s)$$

5. **策略梯度更新**：
   $$\nabla_\theta J = \mathbb{E}_{s \sim \pi_\theta}[\nabla_\theta \log \pi_\theta(a|s) \cdot A(s, a)]$$

6. **TD学习更新Critic**：
   $$\delta = r + \gamma V(s') - V(s)$$
   $$\theta_c \leftarrow \theta_c + \eta \cdot \delta \cdot \nabla_\theta V(s)$$

**策略流程**：
1. 观察订单簿状态、市场微观结构和库存暴露
2. Actor网络输出最优报价策略（价差/数量）
3. 执行交易后计算即时奖励
4. Critic网络评估状态价值，计算优势函数
5. 策略梯度更新Actor参数，TD更新Critic参数
6. 周期性调整库存目标，避免单边暴露

### Python代码示例

```python
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from collections import deque
import random

class ReplayBuffer:
    """经验回放缓冲区"""
    def __init__(self, capacity: int = 10000):
        self.buffer = deque(maxlen=capacity)
    
    def push(self, state, action, reward, next_state, done):
        self.buffer.append((state, action, reward, next_state, done))
    
    def sample(self, batch_size: int):
        batch = random.sample(self.buffer, min(batch_size, len(self.buffer)))
        states, actions, rewards, next_states, dones = zip(*batch)
        return (np.array(states), np.array(actions), 
                np.array(rewards), np.array(next_states), np.array(dones))

class ActorNetwork(nn.Module):
    """Actor网络 - 输出报价策略"""
    def __init__(self, state_dim: int, action_dim: int, hidden_dim: int = 128):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, action_dim),
            nn.Softmax(dim=-1)
        )
    
    def forward(self, state):
        return self.net(state)

class CriticNetwork(nn.Module):
    """Critic网络 - 估计状态价值"""
    def __init__(self, state_dim: int, hidden_dim: int = 128):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )
    
    def forward(self, state):
        return self.net(state)

class ActorCriticMarketMaker:
    """Actor-Critic做市策略"""
    
    def __init__(self, state_dim: int = 20, action_dim: int = 9):
        self.state_dim = state_dim
        self.action_dim = action_dim  # 0-8: 0=收紧价差,4=中立,8=放宽价差
        
        # 网络
        self.actor = ActorNetwork(state_dim, action_dim)
        self.critic = CriticNetwork(state_dim)
        
        # 优化器
        self.actor_optimizer = optim.Adam(self.actor.parameters(), lr=0.001)
        self.critic_optimizer = optim.Adam(self.critic.parameters(), lr=0.002)
        
        # 经验回放
        self.replay_buffer = ReplayBuffer(capacity=10000)
        
        # 超参数
        self.gamma = 0.99
        self.tau = 0.005  # 软更新参数
        self.batch_size = 64
        self.target_update_interval = 10
        
        # 目标网络
        self.target_critic = CriticNetwork(state_dim)
        self.target_critic.load_state_dict(self.critic.state_dict())
        
        # 库存管理
        self.inventory = 0.0
        self.inventory_target = 0.0
        self.inventory_limit = 100.0
        
        # 学习统计
        self.total_steps = 0
        self.episode_rewards = []
        
    def get_state(self, order_book: Dict, market_data: Dict) -> np.ndarray:
        """构建状态向量"""
        features = []
        
        # 订单簿特征 (bid/ask各5档)
        for i in range(5):
            features.append(order_book['bids'][i]['price'] if i < len(order_book['bids']) else 0)
            features.append(order_book['bids'][i]['volume'] if i < len(order_book['bids']) else 0)
            features.append(order_book['asks'][i]['price'] if i < len(order_book['asks']) else 0)
            features.append(order_book['asks'][i]['volume'] if i < len(order_book['asks']) else 0)
        
        # 价差特征
        best_bid = order_book['bids'][0]['price'] if order_book['bids'] else 0
        best_ask = order_book['asks'][0]['price'] if order_book['asks'] else 0
        spread = (best_ask - best_bid) / ((best_ask + best_bid) / 2)
        features.append(spread)
        
        # 库存暴露
        features.append(self.inventory / self.inventory_limit)
        features.append(abs(self.inventory) / self.inventory_limit)
        
        # 市场波动率
        features.append(market_data.get('realized_vol', 0.02) * 50)
        features.append(market_data.get('order_flow_imbalance', 0) * 10)
        
        # 价格趋势
        features.append(market_data.get('mid_price_change', 0) * 100)
        
        # 时间特征
        features.append(market_data.get('hour_of_day', 9.5) / 12 - 1)  # 归一化
        
        return np.array(features, dtype=np.float32)
    
    def select_action(self, state: np.ndarray, epsilon: float = 0.0) -> int:
        """选择报价动作"""
        state_tensor = torch.FloatTensor(state).unsqueeze(0)
        
        with torch.no_grad():
            probs = self.actor(state_tensor).numpy()[0]
        
        # 探索
        if random.random() < epsilon:
            return random.randint(0, self.action_dim - 1)
        
        # 利用
        return np.argmax(probs)
    
    def compute_reward(self, trade: Dict, mid_price: float, 
                      inventory_penalty: float = 0.001) -> float:
        """计算即时奖励"""
        if trade is None:
            return 0.0
        
        # 价差收益
        spread_profit = trade['price'] - mid_price if trade['side'] == 'buy' else mid_price - trade['price']
        
        # 库存成本
        self.inventory += trade['size'] if trade['side'] == 'buy' else -trade['size']
        inventory_cost = inventory_penalty * self.inventory ** 2
        
        # 逆向选择惩罚（价格不利变动时成交）
        adverse_selection = 0.0
        if trade.get('price_move_after', 0) < 0 and trade['side'] == 'buy':
            adverse_selection = abs(trade['price_move_after']) * trade['size']
        elif trade.get('price_move_after', 0) > 0 and trade['side'] == 'sell':
            adverse_selection = abs(trade['price_move_after']) * trade['size']
        
        reward = spread_profit - inventory_cost - adverse_selection * 0.5
        return reward
    
    def update_networks(self, states, actions, rewards, next_states, dones):
        """更新Actor和Critic网络"""
        # 转换到tensor
        states = torch.FloatTensor(states)
        actions = torch.LongTensor(actions)
        rewards = torch.FloatTensor(rewards)
        next_states = torch.FloatTensor(next_states)
        dones = torch.FloatTensor(1 - dones)
        
        # Critic更新
        with torch.no_grad():
            next_values = self.target_critic(next_states).squeeze()
            target_values = rewards + self.gamma * next_values * dones
        
        current_values = self.critic(states).squeeze()
        critic_loss = nn.MSELoss()(current_values, target_values)
        
        self.critic_optimizer.zero_grad()
        critic_loss.backward()
        torch.nn.utils.clip_grad_norm_(self.critic.parameters(), 0.5)
        self.critic_optimizer.step()
        
        # Actor更新 (策略梯度)
        action_probs = self.actor(states)
        dist = torch.distributions.Categorical(action_probs)
        log_probs = dist.log_prob(actions)
        
        # 优势函数
        advantages = (target_values - current_values.detach()).unsqueeze(1)
        actor_loss = -(log_probs * advantages).mean()
        
        self.actor_optimizer.zero_grad()
        actor_loss.backward()
        torch.nn.utils.clip_grad_norm_(self.actor.parameters(), 0.5)
        self.actor_optimizer.step()
        
        return critic_loss.item(), actor_loss.item()
    
    def soft_update_target(self):
        """软更新目标网络"""
        for target_param, param in zip(self.target_critic.parameters(), 
                                         self.critic.parameters()):
            target_param.data.copy_(self.tau * param.data + 
                                     (1 - self.tau) * target_param.data)
    
    def manage_inventory(self, current_inventory: float) -> float:
        """管理库存目标，返回调整后的inventory_target"""
        # 库存限制检查
        if abs(current_inventory) > self.inventory_limit * 0.9:
            # 接近限制，降低目标
            if current_inventory > 0:
                self.inventory_target = min(0, current_inventory - 20)
            else:
                self.inventory_target = max(0, current_inventory + 20)
        else:
            # 库存正常，逐步回归中性
            self.inventory_target *= 0.95
            
        return self.inventory_target
    
    def get_quote(self, state: np.ndarray, mid_price: float) -> Dict:
        """生成报价"""
        action = self.select_action(state)
        
        # 将动作映射到价差和数量
        spread_multiplier = 0.5 + action * 0.25  # 0.5 ~ 2.5
        base_spread = mid_price * 0.001  # 基础价差0.1%
        
        adjusted_spread = base_spread * spread_multiplier
        
        # 库存调整
        if self.inventory > 0:
            # 多头，减少买入报价
            adjusted_spread *= (1 + self.inventory / self.inventory_limit * 0.3)
        else:
            # 空头，减少卖出报价
            adjusted_spread *= (1 - self.inventory / self.inventory_limit * 0.3)
        
        bid_price = mid_price - adjusted_spread / 2
        ask_price = mid_price + adjusted_spread / 2
        
        # 报价数量
        quote_size = max(1, int(10 / spread_multiplier))
        
        return {
            'bid_price': bid_price,
            'ask_price': ask_price,
            'bid_size': quote_size,
            'ask_size': quote_size,
            'spread': adjusted_spread,
            'action': action
        }
    
    def train_episode(self, order_book_history: List[Dict], 
                     market_data_history: List[Dict], trades: List[Dict]) -> Dict:
        """训练一轮"""
        states, actions, rewards, next_states, dones = [], [], [], [], []
        
        for i in range(len(order_book_history) - 1):
            state = self.get_state(order_book_history[i], market_data_history[i])
            next_state = self.get_state(order_book_history[i+1], market_data_history[i+1])
            
            # 获取动作和奖励
            trade = trades[i] if i < len(trades) else None
            mid_price = (order_book_history[i]['bids'][0]['price'] + 
                        order_book_history[i]['asks'][0]['price']) / 2
            
            action = self.select_action(state)
            reward = self.compute_reward(trade, mid_price)
            done = i == len(order_book_history) - 2
            
            # 存储
            self.replay_buffer.push(state, action, reward, next_state, done)
            
            if len(self.replay_buffer.buffer) >= self.batch_size:
                batch = self.replay_buffer.sample(self.batch_size)
                critic_loss, actor_loss = self.update_networks(*batch)
        
        # 更新目标网络
        if self.total_steps % self.target_update_interval == 0:
            self.soft_update_target()
        
        self.total_steps += 1
        
        return {'avg_reward': np.mean(rewards) if rewards else 0}


# 使用示例
if __name__ == "__main__":
    env = ActorCriticMarketMaker(state_dim=20, action_dim=9)
    
    # 模拟订单簿
    sample_order_book = {
        'bids': [
            {'price': 100.0, 'volume': 100},
            {'price': 99.9, 'volume': 150},
            {'price': 99.8, 'volume': 200},
            {'price': 99.7, 'volume': 250},
            {'price': 99.6, 'volume': 300},
        ],
        'asks': [
            {'price': 100.1, 'volume': 100},
            {'price': 100.2, 'volume': 150},
            {'price': 100.3, 'volume': 200},
            {'price': 100.4, 'volume': 250},
            {'price': 100.5, 'volume': 300},
        ]
    }
    
    sample_market = {
        'realized_vol': 0.02,
        'order_flow_imbalance': 0.1,
        'mid_price_change': 0.001,
        'hour_of_day': 10.5
    }
    
    state = env.get_state(sample_order_book, sample_market)
    print(f"状态向量维度: {len(state)}")
    
    quote = env.get_quote(state, mid_price=100.05)
    print(f"报价: 买入{quote['bid_price']:.2f} x {quote['bid_size']}, 卖出{quote['ask_price']:.2f} x {quote['ask_size']}")
    print(f"价差: {quote['spread']:.4f}")
